README
DBIx-FileStore

Perl module that provides tools to insert, delete and manipulate
files stored in a mysql database. 

The implementation and design details are explained throughout 
the module's documentation; in particular see the 
documentation for DBIx::FileStore.

In a nutshell you can copy files into and out of the database 
(respectively) using the commands 'fdbput' and 'fdbget'.

'fdbmv' and 'fdbrm' are also provided, along
with a few other utilities in the bin/ directory.

See the QUICKSTART section below if you just want to get started.

The important things to note are: 
 
A) All the filenames in the filestore are in one flat address space.
   That is, a file like '/a/b/c.txt' is called just that, it's not
   in a folder called /a/b/ (although fdbls does offer some support
   for directory-like viewing of files with common prefixes).

B) The filenames in the filestore cannot contain spaces. 
   Spaces in filenames suck. Respectfully. If you disagree and 
   want me to add support for spaces in filenames, let me 
   know. 

For usage details, see the QUICKSTART section of this file (below)
and the documentation for the filestore scripts in the bin/ 
directory.

For technical details, see the documentation for DBIx::FileStore
and for the utilities in bin/ as well.

There's a video demo showing use of the DBIx::FileStore command
line utilities available at --> http://joshr.com/src/DBIx-FileStore/ 
that is provided in multiple formats. 

INSTALLATION 

See the next section for use of the DBIx::FileStore system.
To install DBIx::Filestore, follow the normal perl test 
and install process:
  
    perl Makefile.PL  (install any prerequisites it might flag)
    make
    make test
    sudo make install
  
QUICKSTART

Here we give a quickstart guide. We assume you've installed 
DBIx::FileStore as in the previous section. To get the filestore
set up: 
  
  1) Have Mysql installed and running. You'll also need DBI and  
     DBD::mysql.
  
  2) Choose what database, dbuser and dbpasswd you're going to use 
     to access mysql.  (DBIx-Filestore uses these to create tables
     called 'files' and 'fileblocks' which are created in step 4. 
     See table-definitions.sql and the documentation for 
     DBIx::FileStore for technical details.)
  
  3) Copy the file fdb.conf-sample to either /etc/fdb.conf or ~/.fdbrc, 
     and edit it to specify your dbname, dbuser and dbpasswd.
  
  4) Create the the needed tables in mysql, in the database you 
     specified above, via a command like:
         % mysql dbname < table-definitions.sql
  
  5) You now have the DBIx-FileStore system installed.  
     It's that simple!
  
  6) As a basic test, let's use the filestore scripts to put 
     a file into the filestore. 
  
         % cd /etc/
         % fdbput hosts
  
     That copies the file '/etc/hosts' into the filestore 
     under the name 'hosts'.  You can use the --local
     option to give files in the filestore different names
     than they have on the filesystem. (See the documentation for
     fdbput and the other scripts for details about their 
     command line options).

     You can see it in the database with fdbls:
  
         % fdbls
         hosts
  
     fdbls also has a few interesting options, including -h, -l, and 
     -m, which show the file's size (formatted for humans), 
     update time (in the filestore), and MD5 checksum, respectively:
  
         % fdbls -h -l -m hosts
         7.44K 20101116 19:28 fp911CO8MLqOLmCGtmx7wA hosts

     You can view a file's contents via the command fdbcat:
  
         % fdbcat hosts | head -3
         127.0.0.1   localhost
         255.255.255.255 broadcasthost
         ::1             localhost 
  
     We can also see some data about the files in the filestore:
  
         % fdbstat
         fdbstat: DB has 7623 bytes in 1 files stored in 1 blocks.

     And you can use 'fdbrm' to remove files from the filestore:
    
         % fdbrm hosts
  
For more info, see the documentation for the 'fdb*' tools:
fdbcat, fdbget, fdbls, fdbmv, fdbput, fdbrm, fdbstat, 
fdbtidy, and fdbslurp.

CAVEAT

"What!?" you say. "You're not supposed to store files directly
in a database!"

And I'd say, "Yes, friend, you're right. But sometimes you just 
want to anyway." I actually wrote this just to see how well it 
would work.

The answer is that it works fine! Modulus the fact that you're
storing large blobs (up to 500K) in a database. (In particular, 
it might make your mysql transaction logs large, if you have
them enabled...)
     
LICENSE AND COPYRIGHT

Copyright (C) 2010-2014 Josh Rabinowitz

This program is free software; you can redistribute it and/or modify it
under the terms of either: the GNU General Public License as published
by the Free Software Foundation; or the Artistic License.

See http://dev.perl.org/licenses/ for more information.



Changes
Revision history for DBIx-FileStore

0.28    2015/01/03
        Fix date test to pass in 2015
        Update docs a little
        Update copyrights to 2015

0.27    2014/12/22
        Test release

0.26    2014/04/22
        Note that fdbslurp is implemented.

0.25    2013/10/08
        Reorganize methods in DBIx::FileStore, document changes.
        Improve docs.

0.24    2013/09/30  
        Improve docs.

0.23    2013/09/30  
        Improve documentation and use of pod.
        Update copyrights.

0.22    2012/12/26  11:39AM
        List File::Slurp, which is used in tests, in prereqs.

0.21    2012/12/23  1:14AM
        Updated MANIFEST
        Re-released so we can tag the code correctly.

0.20    2012/12/23  1:05AM
        Implemented fdbslurp, documented it as implemented.
        Don't read file twice when inserting data any more.
        Code factorings to make fdbslurp work.
        Changes to tests.

0.19    2012/12/23  6:00PM
        Note that scripts are in the bin/ directory, not the scripts/ 
        directory, in the README.
        Improve formatting of POD.

0.18    2012/12/23  5:36PM
        Update copyright years.
        Don't put mkmanifest script into distro.

0.17    2012/12/23  5:25PM
        Move scripts to where build systems expect, from script/ to bin/
        Test that self-tests can _find_ scripts, too.

0.16    2012/12/23  5:10PM
        Fix version numbers to match everywhere.

0.15    2012/12/23  5:10PM
        Use Dist::Zilla in place of Makefile.PL directly.
        Update test to use new location of fdb* scripts.

0.14    -
        No such release.

0.13    2010/12/12  4:15PM
        Don't distribute mkmanifest.
        Added t/actualtest.t, which tests putting files
        in the filestore.
        Added t/exes.t test, which checks that all
        the scripts at least compile.
        README and DBIx::FileStore documentation improved,
        reordered, and rewritten.
        Don't lock database tables by default during changes.
        
0.12    2010/11/28  3:01PM
        Factor more code from command line tools to 
          DBIx::FileStore methods.
        Check complet content md5 when reading data from database.
        Changes to test code and documentation.

0.11    2010/11/24  3:41PM
        Revert default block size to 512K.

0.10    2010/11/24  1:30PM
        Document dependency on DBD::mysql in Makefile.PL.
        Add Usage() function to fdbstat.
        Allow user to set dbhost through the config file.
        Code comments and documentation changes and updates.
        Code cleanups.
        Don't allow filenames to be stored without a name.

0.09    2010/11/19  11:10AM
        Integrate QUICKSTART.pod into README for search.cpan.org.
        Update README, TODO, and MANIFEST.
        Documentation changes and updates.
        Tweak to DBIx::FileStore::ConfigFile for perl 5.6.2.

0.08    2010/11/17  3:02PM
        Change QUICKSTART.txt to QUICKSTART.pod in MANIFEST.
        
0.07    2010/11/17  2:44PM
        Documentation added for mkmanifest.
        Tweaks to POD after seeing DBIx-FileStore 0.05 on CPAN.
        Renamed QUICKSTART.txt to QUICKSTART.pod.

0.06    2010/11/17  1:28PM
        Documentation added for script/fdbstat.
        Incomplete script/fdbslurp script cleaned up
        and documented.
        Comments and documentation cleaned up.
        Mention mysql's max_allowed_packet setting in 
        DBIx/FileStore.pm's pod.  
        TODO updated.
        Not released.

0.05    2010/11/17  12:10PM
        Documentation added for all scripts and modules.
        QUICKSTART.txt added.
        README updated.
        Enable t/pod-coverage.t test.
        fdbput and fdbget have --local=filename option added.
        Removed fdbcp.
        Code cleanup.

0.04    2010/11/16  3:12PM
        This release skipped.

0.03    2010/11/10  3:12PM
        Added needed lib/DBIx/FileStore/ConfigFile.pm file.

0.02    2010/11/10  1:30PM
        Added table-definitions.sql to the source.
        Don't hardcode name of files table in fdbmv.
        Better Usage() output for fdbls.

0.01    2010/11/10 12:30PM
        First version, released on an unsuspecting world.



Makefile.PL

# This file was automatically generated by Dist::Zilla::Plugin::MakeMaker v5.029.
use strict;
use warnings;



use ExtUtils::MakeMaker;



my %WriteMakefileArgs = (
  "ABSTRACT" => "Module to store files in a DBI backend",
  "AUTHOR" => "joshr <joshr\@cpan.org>",
  "CONFIGURE_REQUIRES" => {
    "ExtUtils::MakeMaker" => 0
  },
  "DISTNAME" => "DBIx-FileStore",
  "EXE_FILES" => [
    "bin/fdbrm",
    "bin/fdbget",
    "bin/fdbtidy",
    "bin/fdbslurp",
    "bin/fdbput",
    "bin/fdbcat",
    "bin/fdbmv",
    "bin/fdbls",
    "bin/fdbstat"
  ],
  "LICENSE" => "perl",
  "NAME" => "DBIx::FileStore",
  "PREREQ_PM" => {
    "Config::File" => 0,
    "DBI" => 0,
    "Digest::MD5" => 0,
    "File::Copy" => 0,
    "File::Slurp" => 0,
    "File::Spec::Functions" => 0,
    "File::Temp" => 0,
    "List::Util" => 0,
    "Test::More" => 0
  },
  "VERSION" => "0.28",
  "test" => {
    "TESTS" => "t/*.t"
  }
);


my %FallbackPrereqs = (
  "Config::File" => 0,
  "DBI" => 0,
  "Digest::MD5" => 0,
  "ExtUtils::MakeMaker" => 0,
  "File::Copy" => 0,
  "File::Slurp" => 0,
  "File::Spec::Functions" => 0,
  "File::Temp" => 0,
  "List::Util" => 0,
  "Test::More" => 0
);


unless ( eval { ExtUtils::MakeMaker->VERSION(6.63_03) } ) {
  delete $WriteMakefileArgs{TEST_REQUIRES};
  delete $WriteMakefileArgs{BUILD_REQUIRES};
  $WriteMakefileArgs{PREREQ_PM} = \%FallbackPrereqs;
}

delete $WriteMakefileArgs{CONFIGURE_REQUIRES}
  unless eval { ExtUtils::MakeMaker->VERSION(6.52) };

WriteMakefile(%WriteMakefileArgs);





lib/DBIx/FileStore/ConfigFile.pm
package DBIx::FileStore::ConfigFile;
use strict;

use List::Util qw(first);
use File::Spec::Functions qw(catfile); # to concat dirs & filenames portably 
use Config::File;

use DBIx::FileStore::UtilityFunctions qw(get_user_homedir);

use fields qw( vars_hash verbose ); # you can set $obj->{verbose} for debugging

######################################
# my $conf = new DBIx::FileStore::ConfigFile() 
sub new {
    my DBIx::FileStore::ConfigFile $self = shift;
    unless (ref $self) {
        $self = fields::new($self);
    }
    return $self;
}


############################################
# my $hashref = $conf->read_config_file()
#  or
# my $hashref = $conf->read_config_file( "/etc/fdb-alternate.conf" )
sub read_config_file {
    my ($self, $opt_filename) = @_;

    # choose which config file we're going to use
    my $filename = $opt_filename;
    my $user_dotfile = catfile( get_user_homedir(), ".fdbrc" ); # ~/.fdbrc
    my $etc_conffile = "/etc/fdb.conf";
    unless( $filename ) {
        # explicit sub{} like this for perl 5.6.2
        $filename = first( sub { -e }, $user_dotfile, $etc_conffile );
    }

    unless($filename) {
        die ("$0: Can't find config file to open " . 
            "(tried $user_dotfile and $etc_conffile)\n");
    }
    print "$0: reading $filename\n" if $self->{verbose};

    $self->{vars_hash} = Config::File::read_config_file( $filename );

    # sanity check that there's a dbuser passed.
    unless( $self->{vars_hash}->{dbuser} ) {
        warn "$0: Can't find a 'dbuser' setting in $filename\n";
    }

    return $self->{vars_hash};
}

1;

=pod

=head1 NAME

DBIx::FileStore::ConfigFile -- Find and read filestore conf files.

=head1 SYNOPSIS

    my $conf = new DBIx::FileStore::ConfigFile();
    my $hashref = $conf->read_config_file();

    # these are the fields we use, along wth dbpasswd
    print "db: $hashref->{dbname}, user: $hashref->{dbuser}\n";

=head1 DESCRIPTION

Provides interface to read DBIx::FileStore configuration files.

The read_config_file() method reads from the optionally 
passed configuration file, the file .fdbrc in the user's 
home directory , or /etc/fdb.conf, whichever is found first.

=head1 METHODS

=over 4

=item new DBIx::FileStore::ConfigFile();

my $conf = new DBIx::FileStore::ConfigFile();

Returns a new DBIx::FileStore::ConfigFile object.  

=item $conf->read_config_file() 

my $conf_hash = $conf->read_config_file();

my $conf_hash = $conf->read_config_file( $filename )

Returns a hashref with the name/value pairs parsed from 
the configuration file. The settings expected by 
DBIx-Filestore are: dbname, dbuser, and dbpasswd.

If a $filename is passed by the caller, that file is used as 
the configuration file. Otherwise the module uses 
the file .fdbrc in the current user's home directory, 
or /etc/fdb.conf, whichever is found first. 

If no configuration file can be found, the method dies 
with an error message.

=back

=head1 COPYRIGHT

Copyright (c) 2010-2014 Josh Rabinowitz, All Rights Reserved.

=head1 AUTHORS

Josh Rabinowitz

=cut    



lib/DBIx/FileStore.pm
package DBIx::FileStore;
use strict;

# this reads and writes files from the db.  
use DBI;
use Digest::MD5 qw( md5_base64 );
use File::Copy;

use DBIx::FileStore::ConfigFile;

use fields qw(  dbh dbuser dbpasswd 
                dbhost dbname filetable blockstable  blocksize 
                verbose 
                confhash
                uselocks
                );

our $VERSION = '0.28';  # version also mentioned in POD below.

sub new {
    my ($self, %opts) = @_;
    unless (ref $self) {
        $self = fields::new($self);
    }
    if ($opts{verbose}) { $self->{verbose}=1; }

    my $config_reader = new DBIx::FileStore::ConfigFile();
    my $conf = $self->{confhash} = $config_reader->read_config_file();

    # FOR TESTING WITH 1 BYTE BLOCKS
    #my $block_size = 1; # 1 byte blocks (!)
    my $block_size = 500 * 1024;        # 512K blocks
    $self->{blocksize}   = $block_size;

    #   with 900K (or even 600K) blocks, when inserting binary .rpm files, 
    #   we get
    #
    #   DBD::mysql::db do failed: Got a packet bigger than 'max_allowed_packet' bytes.
    #
    #   We think there's some encoding of the binary data going 
    #   that inflates binary data during transmission. 
    #
    

    ###############################################################################
    # By default, WE DON'T USE LOCKS ANY MORE. Like a real filesystem, you might
    # get interspersed or truncated information if the filesystem is
    # being changed while you're reading!
    ###############################################################################
    $self->{uselocks}    = 0;

    $self->{dbuser}      = $conf->{dbuser} || die "$0: no dbuser set\n";
    $self->{dbpasswd}    = $conf->{dbpasswd} || warn "$0: no dbpasswd set\n";    # this could be ok.
    $self->{dbname}      = $conf->{dbname} || die "$0: no dbname set\n";

    # dbhost defaults to 127.0.0.1
    $self->{dbhost}      = $conf->{dbhost} || "127.0.0.1";

    $self->{filetable}   = "files";
    $self->{blockstable} = "fileblocks";

    my $dsn = "DBI:mysql:database=$self->{dbname};host=$self->{dbhost}";

    my %attr  = ( RaiseError => 1, PrintError => 1, AutoCommit => 1 );  # for mysql

    $self->{dbh} = DBI->connect_cached( 
        $dsn, $self->{dbuser}, $self->{dbpasswd}, \%attr);  
    $self->{dbh}->{mysql_auto_reconnect} = 1;   # auto reconnect

    return $self;
}

sub get_all_filenames {
    my ($self) = @_;
    my $files = $self->{dbh}->selectall_arrayref( # lasttime + 0 gives us an int back
                "select name, c_len, c_md5, lasttime+0 from $self->{filetable} 
                where b_num=0 order by name");
    return $files;
}

sub get_filenames_matching_prefix {
    my ($self, $name) = @_;
    my $pattern = $name . "%"; 
    my $files = $self->{dbh}->selectall_arrayref( # lasttime + 0 gives us an int back
        "select name, c_len, c_md5, lasttime+0 from $self->{filetable} 
         where name like ? and b_num=0 order by name", {}, $pattern);
    return $files;
}

sub rename_file {
    my ($self, $from, $to) = @_;
    die "$0: name not ok: $from" unless name_ok($from);
    die "$0: name not ok: $to" unless name_ok($to);
    # renames the rows in the filetable and the blockstable
    my $dbh = $self->{dbh};
    $self->_lock_tables();

    for my $table ( ( $self->{filetable}, $self->{blockstable} ) ) {
        my $sql = "select name from $table where name like ?";
        $sql .= " order by b_num" if $table eq $self->{filetable};

        my $files = $dbh->selectall_arrayref( $sql, {}, $from . " %");
        for my $f (@$files) {
            (my $num = $f->[0]) =~ s/.* //;
            print "$0: Moving $table:$f->[0], (num $num) to '$to $num'...\n" if $self->{verbose};
            $dbh->do("update $table set name=? where name=?", {}, "$to $num", $f->[0]);
        }
    }

    $self->_unlock_tables();
    return 1;
}

sub delete_file {
    my ($self, $name) = @_;
    die "$0: name not ok: $name" unless name_ok($name);

    my $dbh         = $self->{dbh};
    my $filetable   = $self->{filetable};        # probably "files"
    my $blockstable = $self->{blockstable};    # probably "fileblocks"
    for my $table ( ( $filetable, $blockstable ) ) {
        my $rv = int($dbh->do( "delete from $table where name like ?", {}, "$name %" ));
        if($rv) {
            print "$0: $table: deleted $name ($rv blocks)\n" if $self->{verbose};
        } else {
            warn  "$0: no blocks to delete for $table:$name\n" if $self->{verbose};
        }
    }
    return 1;
}


sub copy_blocks_from_db_to_filehandle {
    my ($self, $fdbname, $filehandle) = @_;
    die "$0: name not ok: $fdbname" unless name_ok($fdbname);
    my $print_to_filehandle_callback = sub {    
        # this is a closure, so $fh comes from the surrounding context
        my $content = shift;
        print $filehandle $content;
    };
    # read all our blocks, calling our callback for each one
    my $ret = $self->_read_blocks_from_db( $print_to_filehandle_callback, $fdbname ); 
    return $ret;
}


# reads the content into $pathname, returns the length of the data read.
sub read_from_db {
    my ($self, $pathname, $fdbname) = @_;

    die "$0: name not ok: $fdbname" unless name_ok($fdbname);

    open( my $fh, ">", $pathname) || die "$0: can't open for output: $pathname\n";

    # this is a function used as a callback and called with each chunk of the data 
    # into a temporary file.  $fh stay in context for the function (closure) below.
    my $print_to_file_callback = sub {    
        # this is a closure, so $fh comes from the surrounding context
        my $content = shift;
        print $fh $content;
    };

    # read all our blocks, calling our callback for each one
    my $ret = $self->_read_blocks_from_db( $print_to_file_callback, $fdbname ); 

    # if we fetched *something* into our scoped $fh to the temp file,
    # then copy it to the destination they asked for, and delete
    # the temp file.
    if (!$fh) {  
        warn "$0: not found in FileDB: $fdbname\n";
    } 
    # clear our fh member
    close($fh) if defined($fh);  # this 'close' should cause the associated file to be deleted
    $fh = undef;

    # return number of bytes read.
    return $ret;    
}


# my $bytes_written = $self->write_to_db( $localpathname, $filestorename );
sub write_to_db {
    my ($self, $pathname, $fdbname) = @_;

    die "$0: name not ok: $fdbname" unless name_ok($fdbname);

    open(my $fh, "<" , $pathname) 
        || die "$0: Couldn't open: $pathname\n";
    

    my $total_length = -s $pathname;   # get the length 
    if ($total_length == 0) { warn "$0: warning: writing 0 bytes for $pathname\n"; }

    my $bytecount = $self->write_from_filehandle_to_db( $fh, $fdbname );

    die "$0: file length didn't match written data length for $pathname: $bytecount != $total_length\n"
        if ($bytecount != $total_length);

    close ($fh) || die "$0: couldn't close: $pathname\n";
    return $total_length;
}

sub write_from_filehandle_to_db {
    my ($self, $fh, $fdbname) = @_;

    my $ctx = Digest::MD5->new; 
    my $dbh = $self->{dbh};
    my $filetable = $self->{filetable};
    my $blockstable = $self->{blockstable};
    my $verbose = $self->{verbose};
    my $size = 0;

    $self->_lock_tables();
    for( my ($bytes,$part,$block) = (0,0,"");           # init
    $bytes = read($fh, $block, $self->{blocksize});    # test 
    $part++, $block="" ) {                              # increment
        $ctx->add( $block );
        $size += length( $block );
        my $b_md5 = md5_base64( $block );

        printf("saving from filehandle into '%s $part'\n", $fdbname )
            if($verbose && $part%25==0); 

        my $name = sprintf("%s %05d", $fdbname, $part);
        $dbh->do("replace into $filetable set name=?, c_md5=?, b_md5=?, c_len=?, b_num=?", {}, 
            $name, "?", $b_md5, 0, $part);
        $dbh->do("replace into $blockstable set name=?, block=?", {}, 
            $name, $block);
    }
    print "\n" if $verbose;
    $dbh->do( "update $filetable set c_md5=?, c_len=? where name like ?", {}, $ctx->b64digest, $size, "$fdbname %" );
    $self->_unlock_tables();

    return $size;
}


#  From here below is utility code and implementation details
# returns the length of the data read,
# calls &$callback( $block ) for each block read.
sub _read_blocks_from_db {
    my ($self, $callback, $fdbname) = @_;
        # callback is called on each block, like &$callback( $block )
    die "$0: name not ok: $fdbname" unless name_ok($fdbname);
    my $dbh = $self->{dbh};
    my $verbose = $self->{verbose};
    my $ctx = Digest::MD5->new();
    
    warn "$0: Fetching rows $fdbname" if $verbose;
    $self->_lock_tables();
    my $cmd = "select name, b_md5, c_md5, b_num, c_len from $self->{filetable} where name like ? order by b_num";
    my @params = ( $fdbname . ' %' );
    my $sth = $dbh->prepare($cmd);
    my $rv =  $sth->execute( @params );
    my $rownum = 0;
    my $c_len  = 0;
    my $orig_c_md5;
    while( my $row = $sth->fetchrow_arrayref() ) {
        $c_len  ||= $row->[5];
        unless ($row && defined($row->[0])) {
            warn "$0: Error: bad row returned?";
            next;
        }
        print "$row->[0]\n" if $verbose;
        unless ($row->[0] =~ /\s+(\d+)$/) {
            warn "$0: Skipping block that doesn't match ' [0-9]+\$': $row->[0]";
            next;
        }
        my $name_num = $1;
        $orig_c_md5 ||= $row->[2];

        # check the MD5...
        if ($row->[2] ne $orig_c_md5) { die "$0: Error: Is DB being updated? Bad content md5sum for $row->[0]\n"; }
        die "$0: Error: our count is row num $rownum, but file says $name_num" 
            unless $rownum == $name_num;
        
        my ($block) =  $dbh->selectrow_array("select block from fileblocks where name=?", {}, $row->[0]);
        die "$0: Bad MD5 checksum for $row->[0] ($row->[1] != " . md5_base64( $block ) 
            unless ($row->[1] eq md5_base64( $block ));
        $ctx->add($block);

        &$callback( $block );   # call the callback, and pass it the block!

        $rownum++;
    }
    $self->_unlock_tables();

    my $retrieved_md5 = $ctx->b64digest();
    die "$0: Bad MD5 checksum for $fdbname ($retrieved_md5 != $orig_c_md5)"
        unless ($retrieved_md5 eq $orig_c_md5);
        
    return $c_len;  # for your inspection
}


sub _lock_tables {
    my $self = shift;
    if ($self->{uselocks}) {
        $self->{dbh}->do("lock tables $self->{filetable} write, $self->{blockstable} write");
    }
}
sub _unlock_tables {
    my $self = shift;
    if ($self->{uselocks}) {
        $self->{dbh}->do("unlock tables");
    }
}


# warns and returns 0 if passed filename ending with like '/tmp/file 1'
# else returns 1, ie, that the name is OK
# Note that this is a FUNCTION, not a METHOD
sub name_ok {
    my $file = shift;
    if (!defined($file) || $file eq "" ) {
        warn "$0: Can't use empty filename\n";
        return 0;
    }
    if ($file && $file =~ /\s/) {
        warn "$0: Can't use filedbname containing whitespace\n";
        return 0;
    }
    if (length($file) > 75) {
        warn "$0: Can't use filedbname longer than 75 chars\n";
        return 0;
    }
    return 1;
}

1;

=pod

=head1 NAME

DBIx::FileStore - Module to store files in a DBI backend

=head1 VERSION

Version 0.28

=head1 SYNOPSIS

Ever wanted to store files in a database? 

This code helps you do that.

All the fdb tools in script/ use this library to 
get at file names and contents in the database.

To get started, see the README file (which includes a QUICKSTART
guide) from the DBIx-FileStore distribution.  

This document details the DBIx::FileStore module implementation.

=head1 FILENAME NOTES

The name of the file in the filestore cannot contain spaces.

The maximum length of the name of a file in the filestore
is 75 characters.

You can store files under any name you wish in the filestore. 
The name need not correspond to the original name on the filesystem.

All filenames in the filestore are in one flat address space.
You can use / in filenames, but it does not represent an actual
directory. (Although fdbls has some support for viewing files in the 
filestore as if they were in folders. See the docs on 'fdbls' 
for details.)


=head1 METHODS

=head2 new DBIx::FileStore()

my $filestore = new DBIx::FileStore();

returns a new DBIx::FileStore object

=head2 get_all_filenames()

my $fileinfo_ref = $filestore->get_all_filenames()

Returns a list of references to data about all the files in the
filestore. 

Each row consist of the following columns:
  name, c_len, c_md5, lasttime_as_int

=head2 get_filenames_matching_prefix( $prefix )

my $fileinfo_ref = get_filenames_matching_prefix( $prefix );

Returns a list of references to data about the files in the 
filestore whose name matches the prefix $prefix.

Returns a list of references in the same format as get_all_filenames().

=head2 read_from_db( $filesystem_name, $storage_name);

my $bytecount = $filestore->read_from_db( "filesystemname.txt", "filestorename.txt" );

Copies the file 'filestorename.txt' from the filestore to the file filesystemname.txt
on the local filesystem.

=head2 rename_file( $from, $to );

my $ok = $self->rename_file( $from, $to );

Renames the file in the database from $from to $to.
Returns 1;

=head2 delete_file( $fdbname );

my $ok = $self->delete_file( $fdbname );

Removes data named $filename from the filestore.

=head2 copy_blocks_from_db_to_filehandle()

my $bytecount = $filestore->copy_blocks_from_db_to_filehandle( $fdbname, $fh );

copies blocks from the filehandle $fh into the fdb at the name $fdbname

=head2 _read_blocks_from_db( $callback_function, $fdbname );

my $bytecount = $filestore->_read_blocks_from_db( $callback_function, $fdbname );

** Intended for internal use by this module. ** 

Fetches the blocks from the database for the file stored under $fdbname,
and calls the $callback_function on the data from each one after it is read.

It also confirms that the base64 md5 checksum for each block and the file contents
as a whole are correct. Die()'s with an error if a checksum doesn't match.

If uselocks is set, lock the relevant tables while data is extracted. 

=head2 write_to_db( $localpathname, $filestorename );

my $bytecount = $self->write_to_db( $localpathname, $filestorename );

Copies the file $localpathname from the filesystem to the name
$filestorename in the filestore.

Locks the relevant tables while data is extracted. Locking should probably 
be configurable by the caller.

Returns the number of bytes written. Dies with a message if the source
file could not be read. 

Note that it currently reads the file twice: once to compute the md5 checksum
before insterting it, and a second time to insert the blocks.

=head2 write_from_filehandle_to_db ($fh, $fdbname)

Reads blocks of the appropriate block size from $fb and writes them 
into the fdb under the name $fdbname.
Returns the number of bytes written into the filestore.

=head1 FUNCTIONS

=head2 name_ok( $fdbname )

my $filename_ok = DBIx::FileStore::name_ok( $fdbname )

Checks that the name $fdbname is acceptable for using as a name
in the filestore. Must not contain spaces or be over 75 chars.

=head1 IMPLEMENTATION

The data is stored in the database using two tables: 'files' and 
'fileblocks'.  All meta-data is stored in the 'files' table, 
and the file contents are stored in the 'fileblocks' table.

=head2 fileblocks table

The fileblocks table has only three fields:

=head3 name 

The name of the block, exactly as used in the fileblocks table. 
Always looks like "filename.txt <BLOCKNUMBER>",
for example "filestorename.txt 00000".

=head3 block

The contents of the named block. Each block is currently set
to be 512K.  Care must be taken to use blocks that are 
not larger than mysql buffers can handle (in particular, 
max_allowed_packet).

=head3 lasttime

The timestamp of when this block was inserted into the DB or updated.

=head2 files table

The files table has several fields. There is one row in the files table 
for each row in the fileblocks table-- not one per file (see IMPLEMENTATION 
CAVEATS, below). The fields in the files table are:

=head3 c_len 

Content length. The content length of the complete file (sum of length of all the file's blocks).

=head3 b_num

Block number. The number of the block this row represents. The b_num is repeated as a five
(or more) digit number at the end of the name field (see above). We denormalize
the data like this so we can quickly and easily find blocks by name or block number.

=head3 b_md5

Block md5. The md5 checksum for the block (b is for 'block') represented by this row.
We use base64 encoding (which uses 0-9, a-z, A-Z, and a few other characters)
to represent md5's because it's a little shorter than the hex 
representation. (22 vs. 32 characters)

=head3 c_md5

Content md5. The base64 md5 checksum for the whole file (c is for 'content') represented by this row.

=head3 lasttime

The timestamp of when this row was inserted into the DB or updated.

=head2 See the file 'table-definitions.sql' for more details about 
the db schema used.

=head1 IMPLEMENTATION CAVEATS

DBIx::FileStore is what I would consider production-grade code, 
but the overall wisdom of storing files in blobs in a mysql database
may be questionable (for good reason).  

That having been said, if you have a good reason to do so, as long 
as you understand the repercussions of storing files in 
your mysql database, then this toolkit offers a stable and 
flexible backend for binary data storage, and it works quite nicely. 

If we were to redesign the system, in particular we might reconsider 
having one row in the 'files' table for each block stored in the 
'fileblocks' table.  Perhaps instead, we'd have one entry in 
the 'files' table per file.

In concrete terms, though, the storage overhead of doing it this way
(which only affects files larger than the block size, which defaults
to 512K) is about 100 bytes per block.  Assuming files larger than
512K, and with a conservative average block size of 256K, the extra 
storage overhead of doing it this way is still only about 0.039% 

=head1 AUTHOR

Josh Rabinowitz, C<< <Josh Rabinowitz> >>

=head1 SUPPORT

You should probably read the documentation for the various filestore command-line
tools:

L<fdbcat>, L<fdbget>, L<fdbls>, L<fdbmv>, L<fdbput>, L<fdbrm>, L<fdbslurp>, L<fdbstat>, and L<fdbtidy>.

=over 4

=item * Search CPAN

You can also read the documentation at:

L<http://search.cpan.org/dist/DBIx-FileStore/>

=back

=head1 LICENSE AND COPYRIGHT

Copyright 2010-2015 Josh Rabinowitz.

This program is free software; you can redistribute it and/or modify it
under the terms of either: the GNU General Public License as published
by the Free Software Foundation; or the Artistic License.

See http://dev.perl.org/licenses/ for more information.

=cut

1; # End of DBIx::FileStore

